tools/crm_resource_runtime.c

```diff
@@ -1290,14 +1290,35 @@ cli_resource_restart(pcmk__output_t *out, pe_resource_t *rsc, const char *host,
 
     pe_working_set_t *data_set = NULL;
 
-    if (!resource_is_running_on(rsc, host)) {
-        const char *id = rsc->clone_name?rsc->clone_name:rsc->id;
-        if(host) {
-            out->err(out, "%s is not running on %s and so cannot be restarted", id, host);
+    /* If host is NULL, there's no need to check anything.  A NULL host will
+     * match all hosts.
+     */
+    if (host != NULL) {
+        bool found = false;
+
+        if (pe_rsc_is_clone(rsc->parent)) {
+            for (GList *iter = uber_parent(rsc)->children; iter != NULL; iter = iter->next) {
```

Comment:
With this many differences, I feel like there's either another resource function hiding in here, or that one of our existing functions needs to be modified somehow.  I haven't really thought this through.  I just have a nagging feeling.

Also, I'm starting to doubt that the call to pe_find_resource_with_flags in crm_resource.c does us much good at all.  Perhaps it should be replaced with whatever we're doing here so that rsc is useful from the start, instead of us having to work around its deficiencies later on.  Again, haven't really thought this through very much.

Reply:
It's amusing that rsc->fns->location(), which is what resource_is_running_on() wraps, and rsc->fns->find_rsc(), which is what pe_find_resource_with_flags() wraps, actually don't need to be variant functions since all variants share the same implementation.

We do need pe_find_resource_with_flags() to find the resource originally requested by the user, because that will be in the CIB (cli_resource_restart() needs it for various purposes). The resource we actually need to act on to restart the original resource is a separate thing, and may not be in the CIB (e.g. a clone instance).

The current implementation of resource_is_running_on() is:

* Get a list of all nodes where the resource is active
* Compare the desired node against each node in that list

but I'm wondering if we could just call ->find_rsc() directly (possibly with a new flag for what we need here, if the existing ones don't fit exactly).